### PR TITLE
Dungeon Maker

### PR DESCRIPTION
**Pull request by ThyWoof**
MIGRATES FROM DMPRO:

- Telema Campaign enablement
- Safe dungeon creation settings
- Custom content auto backup

### CODE REVIEW

File: SolastaCommunityExpansion/Models/DungeonMakerContext.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            UpdateGadgetsPlacement();
+            UpdatePropsPlacement();
+        }
+
+        internal static void BackupAndDelete(string path, UserContent userContent)
+        {
+            var backupDirectory = Path.Combine(Main.MOD_FOLDER, BACKUP_FOLDER);
+
+            if (!Directory.Exists(backupDirectory))
```
Comment (PhilPJL) [inline SolastaCommunityExpansion/Models/DungeonMakerContext.cs:22]: Check not required.  CreateDirectory creates if not exists.

File: SolastaCommunityExpansion/Models/DungeonMakerContext.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            {
+                Directory.CreateDirectory(backupDirectory);
+            }
+
+            var title = userContent.Title;
+            var compliantTitle = IOHelper.GetOsCompliantFilename(title);
+            var destinationPath = Path.Combine(backupDirectory, compliantTitle) + "." + DateTime.Now.ToString("yyyy-MM-dd-HH-mm-ss");
+            var backupFiles = Directory.EnumerateFiles(backupDirectory, compliantTitle + "*").ToList();
+
+            backupFiles.Sort();
```
Comment (PhilPJL) [inline SolastaCommunityExpansion/Models/DungeonMakerContext.cs:32]: You could add an OrderBy() before .ToList() instead

File: SolastaCommunityExpansion/Patches/DungeonMaker/DatabaseSelectionModalPatcher.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+using System.Linq;
+using UnityEngine;
+
+namespace SolastaCommunityExpansion.Patches
+{
+    internal static class DatabaseSelectionModalPatcher
+    {
+        [HarmonyPatch(typeof(DatabaseSelectionModal), "SelectMonsterDefinition")]
+        [SuppressMessage("Minor Code Smell", "S101:Types should be named in PascalCase", Justification = "Patch")]
+        internal static class DatabaseSelectionModalBuildMonsters
```
Comment (PhilPJL) [inline SolastaCommunityExpansion/Patches/DungeonMaker/DatabaseSelectionModalPatcher.cs:13]: DatabaseSelectionModal_BuildMonsters ?

File: SolastaCommunityExpansion/Patches/DungeonMaker/ListElementButtonPatcher.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿using HarmonyLib;
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace SolastaCommunityExpansion.Patches.DungeonEditor
+{
+    // better tooltips on DM selected items
+    [HarmonyPatch(typeof(ListElementButton), "BindItem")]
+    [SuppressMessage("Minor Code Smell", "S101:Types should be named in PascalCase", Justification = "Patch")]
+    internal static class ListElementButtonBindItem
```
Comment (PhilPJL) [inline SolastaCommunityExpansion/Patches/DungeonMaker/ListElementButtonPatcher.cs:10]: same

File: SolastaCommunityExpansion/Patches/DungeonMaker/UserLocationPoolManagerPatcher.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
+﻿using HarmonyLib;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Reflection.Emit;
+
+namespace SolastaCommunityExpansion.Patches
```
Comment (PhilPJL) [inline SolastaCommunityExpansion/Patches/DungeonMaker/UserLocationPoolManagerPatcher.cs:7]: Does this work ok with save by location?

File: SolastaCommunityExpansion/Patches/DungeonMaker/ListElementButtonPatcher.cs (hunk 4 above)
Comment (PhilPJL) [reply SolastaCommunityExpansion/Patches/DungeonMaker/ListElementButtonPatcher.cs:10]: I don't mind either way, but I get a warning of 'unnecessary suppression' if it doesn't need suppressing.

**Review by PhilPJL — DISMISSED** [review]

File: SolastaCommunityExpansion/Models/DungeonMakerContext.cs (hunk 1 above)
Comment (ThyWoof) [reply SolastaCommunityExpansion/Models/DungeonMakerContext.cs:22]: ok.

File: SolastaCommunityExpansion/Patches/DungeonMaker/UserLocationPoolManagerPatcher.cs (hunk 5 above)
Comment (ThyWoof) [reply SolastaCommunityExpansion/Patches/DungeonMaker/UserLocationPoolManagerPatcher.cs:7]: different places even though it should had worked if same. This is for locations JSONs.

**Review by PhilPJL — APPROVED** [review]

**Pull request merged**